### PR TITLE
Add subscription mock

### PR DESCRIPTION
**Pull request by Sytten**
Currently this is no way to resolve multiple values like it is the case with subscription.
So this is a proof of concept, I am happy to rework it if need be (and add more tests/documentation).

### CODE REVIEW

**Comment by Sytten** [comment]
@Mike-Gibson I will wait for your feedback

**Comment by Mike-Gibson** [comment]
Hi @Sytten 

Thanks for creating this PR. Subscriptions came up a while ago in https://github.com/Mike-Gibson/mock-apollo-client/issues/10 but then I didn't hear anything since so I assumed there was no demand for it. Great to see it's needed!

The changes look really good - I really like the way you've added it in using the existing `setRequestHandler` method. I had previously wondered how it might work and I like this approach. 

I assume you're using subscriptions in your production code? If so, have you tried it with the changes in this PR? Does it allow you to test your code which uses Apollo subscriptions?

There are a couple of things I'm wondering after reading the code which would be good to get your thoughts on:

### 1. What happens if there are multiple subscriptions to the same query

I could imagine a real world scenario where the same query is subscribed to, but the variables are different (probably also worth considering when there are multiple subscriptions to the same query with the same variables?)

I can see this API lets the user choose what to return when the handler is invoked, which is great, as for the different variables scenario, it could look like
```typescript
const subscriptionA = createMockSubscription();
const subscriptionB = createMockSubscription();
const handler = jest.fn(vars => vars.someVariable === 'a' ? subscriptionA : subscriptionB);
```
I can see a potential problem when the user doesn't write it this way, or the same query with the same variables is executed - if the same mock subscription is used then I think only the last subscription will work?

Maybe Apollo knows if there's an active subscription and returns the existing observable?

Otherwise, if it's down to the link, I thought it might be a simple case of allowing multiple calls to the `MockSubscription.subscribe` method and keeping a list of observers, but then it raises questions such as, what to do if completed has been called and then there's a new subscription?

### 2. Would it be nice to warn the user if they use `MockSubscription` incorrectly

It might be nice to warn the user if they call `next` when the observer is completed for example. Unless this is already handled by the inner observer?

**Comment by Sytten** [comment]
Hi!

I didn't test it yet with our code, but I was going to do that today. I will report once it is done.

1. I like the way you wrote that, this should most likely be the recommended way of writing it. You are correct when you say that only the last one will be executed because it will call `subscribe` on the `MockSubscription` and override the previous observer. We could allow multiple observers but I feel this would deviate from the "real" behaviour of apollo where subscriptions are independent (aka you should create a new mock subscription for every request). Otherwise `closed` is also problematic. So the suggested exemple should look like:
```typescript
const subscriptions = [];
const handler = jest.fn(vars => {
  const sub = createMockSubscription();
  subscriptions.push(sub);
  return sub;
});
```
2. Good idea, I don't think it is. I believe it is just dropped.

**Comment by Sytten** [comment]
Ok I tested it and it works well for our application. So let me know what you want to do concerning the two items so we  can get it merged soon (I happen to need it :P)

**Comment by Mike-Gibson** [comment]
That's great. So I think you're right - if someone is testing something which subscribes to the same query multiple times, then we should leave it up to them to ensure they are using it right. So thinking about that, would it make sense to throw an error if `MockSubscription.subscribe` is called after another observer has been registered? Or is that going cause trouble for people?

Other than that, would be great to update the docs and also to add an integration test or two to check the end-to-end all works out. I'm happy to help out with both.

I'm hoping I have some time this weekend which I can spend on this as well as a couple of other bits, so should be able to get a version out with this all in for you soon.

**Comment by Sytten** [comment]
I will add a warning so the user will be aware, but throwing is a bit too much IMO since this is for testing anyway (sometimes people need to do weird mocking stuff).

I can do the documentation, if you want to do the end-to-end testing so you get a feel for the solution and make sure it works in cases I might not have thought about :)

Cool, for the moment I am using my fork but not ideal.

**Comment by Mike-Gibson** [comment]
Yep, fair point about throwing vs warning. In fact we warn for when a query is client side, here: https://github.com/Mike-Gibson/mock-apollo-client/blob/a4161a646faf36098129288fbb6eb1f4080b39b3/src/mockLink.ts#L13 so something similar would be good.

Yeah, would be great if you could do the docs, and I will take a look at the integration tests

**Comment by Mike-Gibson** [comment]
Hey @Sytten, I've added integration tests and made a tweak to expose an `IMockSubscription` rather than the `MockSubscription` class - it made it slightly easier to use, as it hides the `subscribe` method from the developer, which should only be used internally by `mockLink`. Does that seem reasonable?

While I was writing the integration tests, it seems that once you call `.error` or `.complete` on the mock subscription, that `closed` will be true and subsequently calling `next/error/complete` has no effect. Maybe it's my lack of using subscriptions/observables, but it wasn't immediately obvious that was the case, and it wasn't a great developer-experience when I encountered this, as nothing warned/errored. Do you know if it would be reasonable to check whether `closed` is true in the `MockSubscription` and log a warning/throw an error if that's the case? I think that would give immediate feedback to the developer that they've done something wrong. Can you see any issue with this?

Similarly, maybe it would be good to log a warning/throw an error if the `next/error/complete` methods are called on `MockSubscription` but there is no inner observer, as it means the query hasn't been executed yet, which I think should only be the case when the developer hasn't set up the test properly. What are your thoughts?

**Comment by Sytten** [comment]
The interface looks good to me!

To be honest I didn't dig very deep into the observable within apollo, but again I don't think throwing is a good idea. A warning could be good but we should have a way to disable it if the intent of the test is to test that particular use case (always annoying to have a `console.warn` that you can't disable in tests).

Yes for that use case I think we should at least do a `console.error`. I still think we should add an error log if the subscribe method is called twice.

Let me know if that is OK with you and I can add those logs (still need to write that doc, didn't have the time last week).

**Comment by Mike-Gibson** [comment]
That all sounds good to me. 

Might be good to have unit tests for the mockSubscription warning behaviour? 

And I'll probably update the subscription integration tests to check that there are no warnings logged after that behaviour's been added. 

No worries about not getting around to it yet. If you want a hand with the docs or adding the logging, give me a shout and I'm happy to give it a go.

**Comment by Sytten** [comment]
@Mike-Gibson I added the console logging (with unit tests) and I added a part in the README.
I didn't test the code in the readme, so if you can double check I would appreciate 🙏 
Once you are done, I think it would be ready to merge and release!

**Comment by Mike-Gibson** [comment]
@Sytten Thanks for the changes.

The warning implementation looks great. I tweaked the option name from `logging` to `disableLogging` as I think it makes it clearer that it's enabled by default. 

I updated the code samples in the README - most of the additions were great. There was a minor typo where the mock subscription was returned which I fixed, and I tweaked some of the text - hope the changes look OK.

If all looks good, I'll merge and create a release.

File: README.md [hunk 1]
```diff
@@ -208,44 +209,57 @@ beforeEach(() => {
...
+    mockSubscription.next({ data: { dog: { id: 1, name: 'Rufus', numberOfBarks: 0 } } });
+  });
+  expect(wrapper.text()).toContain('Rufus has barked 0 time(s)');
+});
+
+it('renders updated dog details when new data is received', () => {
+  act(() => {
+    mockSubscription.next({ data: { dog: { id: 1, name: 'Rufus', numberOfBarks: 1 } } });
+  });
+  expect(wrapper.text()).toContain('Rufus has barked 1 time(s)');
```
Comment (Sytten) [inline README.md:232]: Not all framework guarantee the order of tests, might be better if it was in one test with two expects.

**Review by Sytten — COMMENTED** [review]
Looks good to me! Small comment but nothing serious.

Lets merge this!!!

File: README.md (hunk 1 above)
Comment (Mike-Gibson) [reply README.md:232]: Good spot, thanks. I've updated the sample to be one test and two sets of assertions as suggested.

**Comment by Mike-Gibson** [comment]
@Sytten I've merged the PR, and published v1.1.0 to npm. Thanks for all your help!

File: src/mockLink.ts [hunk 2]
```diff
@@ -23,38 +24,45 @@ export class MockLink extends ApolloLink {
...
+    const key = requestToKey(operation.query);
 
-      const handler = this.requestHandlers[key];
+    const handler = this.requestHandlers[key];
 
-      if (!handler) {
-        throw new Error(`Request handler not defined for query: ${print(operation.query)}`);
-      }
+    if (!handler) {
+      throw new Error(`Request handler not defined for query: ${print(operation.query)}`);
```
Comment (mykolavlasov) [inline src/mockLink.ts:33]: Hi @Sytten @Mike-Gibson 
Throwing an error directly from a function body and not from the returned observable is a not backward-compatible change of the API and it was introduced in the minor version of the package(1.1.0). 
Could you please tell me if it was done intentionally? If not, I could contribute the fix for it.
Thanks

Comment (Sytten) [reply src/mockLink.ts:33]: No idea its been too long, I dont work with apollo anymore

**Pull request merged**